### PR TITLE
Fixes #37936 - As a user, I want to invalidate jwt for specific user( UI )

### PR DESCRIPTION
**Pull request by girijaasoni**
Link to API PR: https://github.com/theforeman/foreman/pull/10397

### CODE REVIEW

File: app/controllers/api/v2/users_controller.rb [hunk 1]
```diff
@@ -111,6 +111,23 @@ def update
...
+        Invalidate JWT for a specific user
+      EOS
+
+      def invalidate_jwt
+        @user = User.find(params[:id])
+        if !@user 
+          render :json => { :error => _("User %s does not exist.") % @user.login}
+        elsif  !@user.can?(:edit_users)
+          deny_access N_("User %s does not have permissions to invalidate" % @user.login)
+        else @user.can?(:edit_users) && @user.invalidate_jwt.blank?
```
Comment (ares) [inline app/controllers/api/v2/users_controller.rb]: I think this is not enough. `edit_users` permission may have specific conditions set. This would allow any user that can edit one specific user's profile to invalidate token for all users, introducing a privilege escalation.

If you look at `can?` [definition](https://github.com/theforeman/foreman/blob/2556ac5014b2d31cc94de498f15b3271b2c3fe0e/app/models/user.rb#L181) you can see it also accepts `subject` which should the permission be verified on. We need to be checking that user can `edit_users` on the specific `@user`.

File: app/helpers/users_helper.rb [hunk 2]
```diff
@@ -33,14 +33,22 @@ def user_action_buttons(user, additional_actions = [])
         :method => :post,
         :data => { :no_turbolink => true })
     end
+    
+    if user != User.current
+      additional_actions <<  display_link_if_authorized(_("Invalidate JWT"),
```
Comment (ares) [inline app/helpers/users_helper.rb]: the exact same problem like with `can?` above, the auth object needs to be specified. If this is used on some users listing page, to avoid N+1 issue, an authorized object needs to be used.

**Review by ares — COMMENTED** [review]
Leaving the comment for the the proper review, this can be easily missed but I believe is important to get it right.

File: app/controllers/api/v2/users_controller.rb (hunk 1 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: Ack, I'll look into updating the subject

File: app/helpers/users_helper.rb (hunk 2 above)
Comment (girijaasoni) [reply app/helpers/users_helper.rb]: I am already adding  `:auth_object` here, is [this](https://github.com/theforeman/foreman/pull/10357/commits/cc036ad9ccfd90905d5dd6d22cd29ad926dc4ae5#diff-0a32d4e9f83e8c59487817bce4ccece334edc27596e9c7c9e8e2888ed4b0fc97R39) what you're talking about?

File: app/controllers/api/v2/users_controller.rb [hunk 3]
```diff
@@ -111,6 +111,23 @@ def update
         end
       end
 
+      api :PATCH, "/users/:id/invalidate_jwt", N_("Get vm attributes of host")
```
Comment (ShimShtein) [inline app/controllers/api/v2/users_controller.rb]: Copy-paste leftovers?

File: app/controllers/api/v2/users_controller.rb [hunk 4]
```diff
@@ -111,6 +111,23 @@ def update
...
+      param :id, String, :required => true
```
Comment (ShimShtein) [inline app/controllers/api/v2/users_controller.rb]: I suppose you would want to have an array parameter here. To be able to invalidate tokens for multiple users.

File: app/controllers/api/v2/users_controller.rb [hunk 5]
```diff
@@ -111,6 +111,23 @@ def update
...
+      description <<-DOC
+        Invalidate JWT for a specific user
+      DOC
+
+      def invalidate_jwt
+        @user = User.find(params[:id])
```
Comment (ShimShtein) [inline app/controllers/api/v2/users_controller.rb]:
```suggestion
        @user = User.where(id: params[:id]).authorized(:edit_user)
```
This will filter only the users that `edit_user` permission applies to them, including custom filters.

Now you will have a set of users that you can safely delete the tokens for:
``` ruby
JwtSecret.where(user_id: @users).destroy_all
```

File: app/controllers/api/v2/users_controller.rb [hunk 6]
```diff
@@ -12,8 +12,8 @@ class UsersController < V2::BaseController
         ['compute_attributes']
 
       before_action :find_optional_nested_object
-      skip_before_action :authorize, :only => [:extlogin]
-      before_action :authenticate, :only => [:extlogin]
+      skip_before_action :authorize, :only => [:extlogin, :invalidate_jwt]
```
Comment (ShimShtein) [inline app/controllers/api/v2/users_controller.rb]: Why do you need to skip authorization for `invalidate_jwt` action?
You can specify the action in the security block: https://github.com/theforeman/foreman/blob/eb66c09e63ba9b04f3d86c240ec25de52647bc84/config/initializers/f_foreman_permissions.rb#L566
This will make sure the current user has the `edit_users` permission before accessing the endpoint.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -8,7 +8,7 @@ class UsersController < ApplicationController
   rescue_from ActionController::InvalidAuthenticityToken, with: :login_token_reload
   skip_before_action :require_mail, :only => [:edit, :update, :logout, :stop_impersonation]
   skip_before_action :require_login, :check_user_enabled, :authorize, :session_expiry, :update_activity_time, :set_taxonomy, :set_gettext_locale_db, :only => [:login, :logout, :extlogout]
-  skip_before_action :authorize, :only => [:extlogin, :impersonate, :stop_impersonation]
+  skip_before_action :authorize, :only => [:extlogin, :impersonate, :stop_impersonation, :invalidate_jwt]
```
Comment (ShimShtein) [inline app/controllers/users_controller.rb]: The same as for the API endpoint. You can specify the new action in the security block:
https://github.com/theforeman/foreman/blob/eb66c09e63ba9b04f3d86c240ec25de52647bc84/config/initializers/f_foreman_permissions.rb#L565

**Comment by MariaAga** [comment]
Tested and does as needed, one small thing that acts weird to me is that a user cant invalidate their own jwt, so for example admin doesnt have the invalidate button, but a user with edit_users permission can invalidate admins tokens.

**Comment by girijaasoni** [comment]
> Tested and does as needed, one small thing that acts weird to me is that a user cant invalidate their own jwt, so for example admin doesnt have the invalidate button, but a user with edit_users permission can invalidate admins tokens.

It's because there will be a different button in the user details page for this, from UI Perspective, for a user to invalidate tokens for themselves, they will have it in the profile page

File: app/controllers/api/v2/users_controller.rb (hunk 3 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: yes, my bad

File: app/controllers/api/v2/users_controller.rb (hunk 4 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: currently in the scope, (atleast UI wise) we can invalidate tokens for a specific user or all users because otherwise I would have to add a select action in the UI. I was thinking, I can add another endpoint to invalidate all at once. Allowing multiple (but not all) users token invalidation can take some time to integrate

**Comment by MariaAga** [comment]
@girijaasoni I think it looks a bit weird in the UI that other users have the invalidate button, but not the current (admin) user: 
admin:
Screenshot from 2024-10-29 11-54-26
other user:
Screenshot from 2024-10-29 11-53-41

**Comment by girijaasoni** [comment]
> @girijaasoni I think it looks a bit weird in the UI that other users have the invalidate button, but not the current (admin) 
Some users don't have `edit_users` permissions and they shall still be able to invalidate tokens for themselves which is why it was suggested that we have the invalidate option in the user details page if the user wants to do it for self.

**Comment by girijaasoni** [comment]
Test failures seem unrelated

File: app/controllers/api/v2/users_controller.rb [hunk 8]
```diff
@@ -111,6 +112,25 @@ def update
         end
       end
 
+      api :PATCH, "/users/:id/invalidate_jwt", N_("Invalidate JWT for a specific userS")
+      description <<-DOC
```
Comment (stejskalleos) [inline app/controllers/api/v2/users_controller.rb]: @Lennonka review please of the text & labels

File: app/controllers/api/v2/users_controller.rb [hunk 9]
```diff
@@ -111,6 +112,25 @@ def update
...
+        Invalidate JWT for a specific users
```
Comment (stejskalleos) [inline app/controllers/api/v2/users_controller.rb]: If the description is one-liner, then I would go with `description 'Invalidate JWT for a specific users'`

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -93,6 +93,16 @@ def impersonate
     end
   end
 
+  def invalidate_jwt
+    @user = find_resource(:edit_users)
+    User.find_by(id: @user.id).jwt_secret&.destroy
```
Comment (stejskalleos) [inline app/controllers/users_controller.rb]: Why? You already have the `@user` object

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -93,6 +93,16 @@ def impersonate
...
+    if @user.jwt_secret.blank?
```
Comment (stejskalleos) [inline app/controllers/users_controller.rb]: Do we have to have this if statement?
How can it happen that the `jwt_secret` IS NOT blank?

File: app/controllers/api/v2/users_controller.rb [hunk 12]
```diff
@@ -111,6 +112,26 @@ def update
...
 
+      api :PATCH, "/users/:id/invalidate_jwt", N_("Invalidate JWT for a specific users")
+      description <<-DOC
+        Invalidate JWT for a specific users
+      DOC
+
+      def invalidate_jwt
+        if params[:search].blank?
+          process_resource_error
+        end
```
Comment (stejskalleos) [inline app/controllers/api/v2/users_controller.rb]:
```suggestion
        process_resource_error if params[:search].blank?

```

File: app/controllers/api/v2/users_controller.rb [hunk 13]
```diff
@@ -111,6 +112,26 @@ def update
...
+        @users = resource_scope_for_index(:permission => :edit_users)
+        if @users.empty?()
+          render :json => { :error => _("No record found for '%s'") % params[:search]}
+        else
+          response = JwtSecret.where(user_id: @users).destroy_all
+          if response.blank?
```
Comment (stejskalleos) [inline app/controllers/api/v2/users_controller.rb]: Is the `if response.blank?` needed? In which cases is it not blank?

**Review by stejskalleos — CHANGES_REQUESTED** [review]

**Comment by stejskalleos** [comment]
Failing tests are fixed in https://github.com/theforeman/foreman/pull/10377 and https://github.com/fog/fog-libvirt/pull/163

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (girijaasoni) [reply app/controllers/users_controller.rb]: its to coordinate with the permissions

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (girijaasoni) [reply app/controllers/users_controller.rb]: I mean just like a safety check

File: app/controllers/api/v2/users_controller.rb (hunk 12 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: ack

File: app/controllers/api/v2/users_controller.rb (hunk 13 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: I mean it's optional, just like for safety. I'm open to remove it on vote? what do you think @ShimShtein

File: app/controllers/api/v2/users_controller.rb [hunk 14]
```diff
@@ -111,6 +112,26 @@ def update
         end
       end
 
+      api :PATCH, "/users/:id/invalidate_jwt", N_("Invalidate JWT for a specific users")
```
Comment (Lennonka) [inline app/controllers/api/v2/users_controller.rb]:
```suggestion
      api :PATCH, "/users/:id/invalidate_jwt", N_("Invalidate all JWTs for the user")
```
We're invalidating _all_ JWTs for a _single_ user, right?

File: app/controllers/api/v2/users_controller.rb [hunk 15]
```diff
@@ -111,6 +112,26 @@ def update
...
+      description <<-DOC
+        Invalidate JWT for a specific users
```
Comment (Lennonka) [inline app/controllers/api/v2/users_controller.rb]:
```suggestion
        Invalidate all JSON Web Tokens (JWTs) for a user or users.
        The users you specify will no longer be able to register hosts by using their JWTs.
```

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -93,6 +93,16 @@ def impersonate
...
+      process_success(
+        :success_msg => _('Successfully invalidated JWT for %s.') % @user.login
```
Comment (Lennonka) [inline app/controllers/users_controller.rb]:
```suggestion
        :success_msg => _('Successfully invalidated JWTs for %s.') % @user.login
```
We're invalidating _all_ of user's JWTs, right?

File: app/helpers/users_helper.rb [hunk 17]
```diff
@@ -34,6 +34,13 @@ def user_action_buttons(user, additional_actions = [])
         :data => { :no_turbolink => true })
     end
 
+    if user != User.current
+      additional_actions << display_link_if_authorized(_("Invalidate JWT"),
```
Comment (Lennonka) [inline app/helpers/users_helper.rb]:
```suggestion
      additional_actions << display_link_if_authorized(_("Invalidate JWTs"),
```

File: app/controllers/api/v2/users_controller.rb [hunk 18]
```diff
@@ -111,6 +112,26 @@ def update
...
+        if params[:search].blank?
+          process_resource_error
+        end
+        @users = resource_scope_for_index(:permission => :edit_users)
+        if @users.empty?()
+          render :json => { :error => _("No record found for '%s'") % params[:search]}
+        else
+          response = JwtSecret.where(user_id: @users).destroy_all
+          if response.blank?
+            process_success _('Successfully invalidated JWT for %s.' % @users.pluck(:login).to_sentence)
```
Comment (Lennonka) [inline app/controllers/api/v2/users_controller.rb]:
```suggestion
            process_success _('Successfully invalidated JWTs for %s.' % @users.pluck(:login).to_sentence)
```

**Review by Lennonka — CHANGES_REQUESTED** [review]
My suggestions for the microcopy. It would be great to get the singulars and plurals right.

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (stejskalleos) [reply app/controllers/users_controller.rb]: I don't understand. You are checking permissions on the controller action, right?
If users don't have `edit_users` permission, they cannot use the  `invalidate_jwt` action.
So why to recheck it here?

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (stejskalleos) [reply app/controllers/users_controller.rb]: > I mean just like a safety check

Yes, but is it really needed?
How the `User.find_by(id: @user.id).jwt_secret&.destroy` can result in `@user.jwt_secret` not being blank, that's what I'm asking.

File: app/controllers/api/v2/users_controller.rb (hunk 13 above)
Comment (stejskalleos) [reply app/controllers/api/v2/users_controller.rb]: Well, I would like to see the reasoning behind the check.
I want to know which scenario the response is not blank in.

File: app/controllers/api/v2/users_controller.rb (hunk 14 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: for API, we are doing it for multiple users

File: app/controllers/api/v2/users_controller.rb (hunk 13 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: I don't think there would be a scenario as such. 
I can remove the if condition

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (girijaasoni) [reply app/controllers/users_controller.rb]: This is the only place I'm checking permissions in. No other place

File: app/controllers/api/v2/users_controller.rb (hunk 14 above)
Comment (Lennonka) [reply app/controllers/api/v2/users_controller.rb]: But the user is identified by an ID that must be unique for each user. So how is it for multiple users?

Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: for API, we are using the scoped search approach, so the search params in the curl command would be something like `{"seach" : "id ^ (1,2,3)"}` translating to search for id in 1, 2 ,3.

Comment (Lennonka) [reply app/controllers/api/v2/users_controller.rb]: Okay. In that case I suggest:
```suggestion
      api :PATCH, "/users/:id/invalidate_jwt", N_("Invalidate JWT for specific users")
```

Comment (stejskalleos) [reply app/controllers/api/v2/users_controller.rb]: If you use the search param to get users, then there should not be any `:id` in the URL.
`api :PATCH, "/users/invalidate_jwt"`

File: app/controllers/api/v2/users_controller.rb [hunk 19]
```diff
@@ -111,6 +112,22 @@ def update
...
       end
 
+      api :PATCH, "/users/invalidate_jwt", N_("Invalidate all JSON Web Tokens (JWTs) for a user or users.")
+      description <<-DOC
+      The users you specify will no longer be able to register hosts by using their JWTs.
+      DOC
+
+      def invalidate_jwt
+        process_resource_error if params[:search].blank?
+        @users = resource_scope_for_index(:permission => :edit_users)
```
Comment (ShimShtein) [inline app/controllers/api/v2/users_controller.rb]: I think we need to add `except_hidden` scope to here too

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -93,6 +93,14 @@ def impersonate
     end
   end
 
+  def invalidate_jwt
+    @user = find_resource(:edit_users)
+    User.find_by(id: @user.id).jwt_secret&.destroy
```
Comment (ShimShtein) [inline app/controllers/users_controller.rb]: I still think that 
```suggestion
    find_resource(:edit_users).jwt_secret&.destroy
```
will do the trick

**Review by ShimShtein — CHANGES_REQUESTED** [review]

File: app/helpers/users_helper.rb [hunk 21]
```diff
@@ -34,6 +34,13 @@ def user_action_buttons(user, additional_actions = [])
         :data => { :no_turbolink => true })
     end
 
+    if user != User.current
+      additional_actions << display_link_if_authorized(_("Invalidate JWTs"),
+        hash_for_invalidate_jwt_user_path(:id => user.id).merge(:auth_object => user, :permission => "edit_users"),
+        :method => :patch, :id => user.id,
+        :data => { :confirm => _("Invalidate tokens for %s?") % user.name })
```
Comment (Lennonka) [inline app/helpers/users_helper.rb]:
```suggestion
        :data => { :confirm => _("Invalidate all JSON Web Tokens for %s?") % user.name })
```
I strongly suggest using "JWTs" here or spelling them out because "tokens" might get confused with Personal Access Tokens.

File: app/controllers/api/v2/users_controller.rb [hunk 22]
```diff
@@ -111,6 +112,22 @@ def update
         end
       end
 
+      api :PATCH, "/users/invalidate_jwt", N_("Invalidate all JSON Web Tokens (JWTs) for a user or users.")
```
Comment (ekohl) [inline app/controllers/api/v2/users_controller.rb]: This now looks like it invalidates it for all users and there's no way to specify it for users. REST API wise I think this is also poor design. If you have a user that's called `invalidate_jwt` then it overlaps with other resources.

I'd expect `DELETE /users/:id/jwt` to delete all JWTs for a specific user while for all JWTs it should operate on the JWT collection, so something like `DELETE /jwt`. We currently don't have such an endpoint so you'd need to create that.

File: app/controllers/api/v2/users_controller.rb [hunk 23]
```diff
@@ -96,7 +97,7 @@ def create
       api :PUT, "/users/:id/", N_("Update a user")
       description <<-DOC
         Adds role 'Default role' to the user if it is not already present.
-        Only another admin can change the admin account attribute.
+      Only another admin can change the admin account attribute.
```
Comment (ekohl) [inline app/controllers/api/v2/users_controller.rb]: Please keep correct indenting.
```suggestion
        Only another admin can change the admin account attribute.
```

**Review by ekohl — DISMISSED** [review]
Please add these to `PersonalAccessTokensController`, not the users controller. That's already where we have all the resources to manage them.

https://restful-api-design.readthedocs.io/en/latest/ is an old document, but it's where I learned RESTful API design. Please read https://restful-api-design.readthedocs.io/en/latest/urls.html#url-structure to understand how to think about collections and resources. https://restful-api-design.readthedocs.io/en/latest/methods.html is another on the methods to use. It doesn't explicitly mention `DELETE` on a (sub)collection, but that doesn't mean it can't be done. I'd argue it's well defined.

File: config/initializers/f_foreman_permissions.rb [hunk 24]
```diff
@@ -562,8 +562,8 @@
       :users => [:new, :create],
       :"api/v2/users" => [:create]
     map.permission :edit_users,
-      :users => [:edit, :update],
-      :"api/v2/users" => [:update]
+      :users => [:edit, :update, :invalidate_jwt],
+      :"api/v2/users" => [:update, :invalidate_jwt]
```
Comment (ekohl) [inline config/initializers/f_foreman_permissions.rb]: Where does `invalidate_jwt` come from? I don't see it defined in `PermissionsList`

File: app/controllers/api/v2/users_controller.rb (hunk 22 above)
Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: I think a user with `invalidate_jwt` would be an edge case
we are using scoped search approach in the API request so the user can use id or name or login anything for reference. From API POV, its valid for one specific user or multiple users as well. Request params would go like "{"search" : "id ^ (1, 2, 3)" } , does it make sense?

Comment (girijaasoni) [reply app/controllers/api/v2/users_controller.rb]: cc @ShimShtein

File: config/initializers/f_foreman_permissions.rb (hunk 24 above)
Comment (girijaasoni) [reply config/initializers/f_foreman_permissions.rb]: it's an action in userscontrollers, do we need to add it in the permissions list?

**Comment by girijaasoni** [comment]
As per further discussion, we have decided to split the UI and the API PR. @ekohl suggests we add a separate controller for JWTs. It will be updated in the next PR 
CC: @ShimShtein

**Review by ShimShtein — DISMISSED** [review]

**Comment by ShimShtein** [comment]
@ekohl This one looks good. Feel free to merge, if it's OK with you

**Review by Lennonka — APPROVED** [review]
Microcopy looks good to me.

File: app/helpers/users_helper.rb [hunk 25]
```diff
@@ -34,6 +34,13 @@ def user_action_buttons(user, additional_actions = [])
         :data => { :no_turbolink => true })
     end
 
+    if user != User.current
```
Comment (stejskalleos) [inline app/helpers/users_helper.rb:37]: With this condition, admin users cannot invalidate tokens for other users.

**Review by stejskalleos — CHANGES_REQUESTED** [review]
:green_apple: CI
:green_apple: Tests
:green_apple: Code

One scenario doesn't work for me, and the other is to think about.

1 - Users with the `Viewer` role cannot invalidate tokens for themselves.
2 - If users don't have `view|edit _users` permissions, how can they invalidate their own tokens? There are no actions at `My Account` page.

**Comment by girijaasoni** [comment]
> 🍏 CI 🍏 Tests 🍏 Code
> 
> One scenario doesn't work for me, and the other is to think about.
> 
> 1 - Users with the `Viewer` role cannot invalidate tokens for themselves. 2 - If users don't have `view|edit _users` permissions, how can they invalidate their own tokens? There are no actions at `My Account` page.

This PR Is for the story "User to invalidate tokens for specific(other) users" , I will be updating invalidating self in the next PR as its a separate story

File: test/controllers/users_controller_test.rb [hunk 26]
```diff
@@ -158,6 +158,33 @@ class UsersControllerTest < ActionController::TestCase
     assert !User.exists?(user.id)
   end
 
+  test "Admin should be able to invalidate jwt for any user" do
+    User.current = users(:admin)
+    user = users(:two)
+    FactoryBot.build(:jwt_secret, token: 'test_jwt_secret', user: user)
+    patch :invalidate_jwt, params: { :id => user.id }, session: set_session_user
+    user.reload
```
Comment (stejskalleos) [inline test/controllers/users_controller_test.rb:166]:
```suggestion
```

Why is this needed? (in all tests)

File: test/controllers/users_controller_test.rb [hunk 27]
```diff
@@ -158,6 +158,33 @@ class UsersControllerTest < ActionController::TestCase
...
+    assert_response :found
+  end
+
+  test 'user with edit users permission should be able to invalidate jwt for another user' do
+    setup_user 'edit', 'users'
+    user = users(:two)
+    FactoryBot.build(:jwt_secret, token: 'test_jwt_secret', user: user)
+    patch :invalidate_jwt, params: { :id => user.id }, session: set_session_user
```
Comment (stejskalleos) [inline test/controllers/users_controller_test.rb]: `set_session_user` use `:admin` user as default. The test is using `admin` user instead of "non-admin"

**Review by stejskalleos — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (nofaralfasi) [reply app/controllers/users_controller.rb]: Each user can have only one JWT. In my opinion, we should keep it singular.

File: test/controllers/users_controller_test.rb (hunk 26 above)
Comment (girijaasoni) [reply test/controllers/users_controller_test.rb:166]: it needs to reload so JWT is updated

Comment (stejskalleos) [reply test/controllers/users_controller_test.rb:166]: That doesn't make sense.
After you call `patch`, the `response` is already set. If you remove the `user.reload`, test will work same way. At least on my setup.

**Review by stejskalleos — APPROVED** [review]
:green_apple: **LGTM**, failing tests are not related.
Let's get this in.

**Comment by stejskalleos** [comment]
Thanks @girijaasoni and everyone involved.
